### PR TITLE
Fixes #24622 - UEFI HTTP boot loader options

### PR DESCRIPTION
**Pull request by lzap**
UEFI clients are capable of booting from HTTP or HTTPS if they are given a valid URL via DHCP filename option.

Let's create such PXE Loader options. The only difference from existing options is that it must contain hostname of the boot server. By default this will be TFTP Smart Proxy if set, Foreman unattended_url otherwise.

For better flexibility, two new provisioning settings will be introduced: port number and root path. This allows users easily run their own web server exposing TFTP directory with grub2 configuration files over HTTP on arbitrary root path.

### CODE REVIEW

File: app/assets/javascripts/host_edit.js
Comment (lzap) [inline app/assets/javascripts/host_edit.js]: This warning appears if you select any of the HTTP UEFI boot options, I just want to make clear that it will work if you do some manual changes (e.g. set up HTTP server on TFTP smart proxy).

File: app/models/concerns/pxe_loader_support.rb
Comment (lzap) [inline app/models/concerns/pxe_loader_support.rb]: I also add iPXE UEFI which simply loads `ipxe-x64.efi` which can be grabbed from here: http://boot.ipxe.org/

We don't ship this with Foreman or Satellite tho, users can still try this workflow if they like it.

File: app/models/concerns/pxe_loader_support.rb
Comment (lzap) [inline app/models/concerns/pxe_loader_support.rb:9]: I am cleaning those regexps here, covered with many tests.

File: app/models/setting/provisioning.rb
Comment (lzap) [inline app/models/setting/provisioning.rb]: New settings for better flexibility, all the rest are tests.

File: app/models/concerns/pxe_loader_support.rb
Comment (stbenjam) [reply app/models/concerns/pxe_loader_support.rb]: Just my initial impression after a quick skim: I have mixed feelings about using the tftp_proxy for building the URL.   

Back when the templates proxy was implemented, this approach was taken and it turned out to be problematic.  I realize this is a different use case, but I think the end goal would be to have a smart proxy plugin called HTTPBOOT, right? Should it be responsible for knowing the URL for HTTP and HTTPS boot and returning that to Foreman like template_url?

Alos, I realize this is the first "manual" approach.  The user is going to have to configure HTTP to serve the files.  Could we, at least for the initial deployment, use host parameters instead for the URL? Then once we implement a smart proxy plugin that's a little more intelligent use that.

Comment (lzap) [reply app/models/concerns/pxe_loader_support.rb]: If you prefer having our own module, I can create one, let's do this properly from the day one. The only concern is security bugs, allowing others to read arbitrary files. History shows that many people failed, many times :-)

But yes, this is cleaner from the user and dev perspective, the only drawback is that we will have a feature "httpboot" dependent on other feature "tftp" and I am not sure if smart-proxy plugin API supports that.

**Comment by lzap** [comment]
New version is up, it uses new feature called HTTPBoot, make sure to pick it. If it's not picked, it falls back to Foreman unattended_url setting. Smart proxy with enabled "httpboot" module is required. I haven't tested this end-to-and yet.

File: app/assets/javascripts/host_edit.js
Comment (stbenjam) [reply app/assets/javascripts/host_edit.js]: Nitpick: needs a space between `boot.` and `Manual`

File: app/models/subnet.rb
Comment (stbenjam) [inline app/models/subnet.rb:57]: There's a `def proxies` method in Subnet, it should return `httpboot` too:

https://github.com/theforeman/foreman/blob/develop/app/models/subnet.rb#L213

**Review by stbenjam — COMMENTED** [review]
I successfully end-to-end tested this and the smart-proxy PR.   I left some notes on the smart proxy PR but other than minor issues, this part looks good.

**Comment by lzap** [comment]
Done.

**Review by stbenjam — APPROVED** [review]
Code climate issue is related to PR.  But otherwise, this looks good to me.

**Comment by stbenjam** [comment]
Oh my checkmark is green.  I'm not really a Foreman committer though, I think I only have access to the repo because of my membership in the packaging groups.  You'll need someone else to review, too.

**Comment by lzap** [comment]
Thanks for testing this, some can merge this @timogoebel, @ares or @tbrisker ? There is a linked smart-proxy PR as well.

**Review by timogoebel — APPROVED** [review]
Looks good to me.

**Comment by timogoebel** [comment]
Thanks, @lzap.

**Pull request merged**
